wandb/data_types.py

```diff
@@ -748,6 +748,107 @@ def json_helper(val):
 
         return json_dict
 
+    def iterrows(self):
+        """Iterate over rows as (ndx, row)
+        Yields
+        ------
+        index : int
+            The index of the row.
+        row : List[any]
+            The data of the row
+        """
+        for ndx in range(len(self.data)):
+            yield ndx, self.data[ndx]
+
+
+class _PartitionTablePartEntry:
+    """Helper class for PartitionTable to track its parts
+    """
+
+    def __init__(self, entry, source_artifact):
+        self.entry = entry
+        self.source_artifact = source_artifact
+        self._part = None
+
+    def get_part(self):
+        if self._part is None:
+            self._part = self.source_artifact.get(self.entry.path)
+        return self._part
+
+    def free(self):
+        self._part = None
+
+
+class PartitionedTable(Media):
+    """ PartitionedTable represents a table which is composed
+    by the union of multiple sub-tables. Currently, PartitionedTable
+    is designed to point to a directory within an artifact.
+    """
+
+    artifact_type = "partitioned-table"
+
+    def __init__(self, parts_path):
+        """
+        Args:
+            parts_path (str): path to a directory of tables in the artifact
+        """
+        super(PartitionedTable, self).__init__()
+        self.parts_path = parts_path
+        self._loaded_part_entries = {}
+
+    def to_json(self, artifact):
+        json_obj = super(PartitionedTable, self).to_json(artifact)
+        json_obj["parts_path"] = self.parts_path
+        return json_obj
+
+    @classmethod
+    def from_json(cls, json_obj, source_artifact):
+        instance = cls(json_obj["parts_path"])
+        entries = source_artifact.manifest.get_entries_in_directory(
+            json_obj["parts_path"]
+        )
+        for entry in entries:
+            instance._add_part_entry(entry, source_artifact)
+        return instance
+
+    def iterrows(self):
```

Comment:
I like this pattern. I feel like there's a straightforward path to adding things like selects, filters, and limits with something like:

```
partition_table
  .select('id', 'val')
  .where('val > 2')
  .limit(10)
  .iterrows()
```

Any reason why you didn't go with the Python `__iter__` and `__next__` approach? It might be easy to just return something like:

```
def __iter__(self):
  class Iterator:
    def __next__(self):
      for entry_path in self._loaded_part_entries:
        ...
        yield ...
    return Iterator()
```
The resulting syntax might be nice: `for idx, row in partition_table ...`

Reply:
Thanks! yeah, this will be really nice as we move to more "read-side" operations. I chose to make an explicit iterator to match the pandas style `iterrows` method which does the same thing. I try to stay away from internally created classes as it makes for difficult type checking in the future (once we type all these files). Also, the current `iterrows` method returns a Generator which is a subclass of an Iterator. Effectively the same thing. Anyhow, I think ultimately looping over the object itself such as `for idx, row in partition_table` is totally possible, but if we make that call now we can never go back.

Reply:
Makes sense. I wasn't aware of the pandas `iterrows`. I think adopting the same terminology makes sense.

Reply:
@annirudh: note how that little snippet of code you wrote above looks a lot like the compute graph "code" in the UI ;)